kube-core/src/conversion.rs

```diff
@@ -0,0 +1,141 @@
+//! Contains utilities for implements custom resource conversion webhooks.
+//! Abstractions in this module and its submodules may be ordered:
+//! - `low_level` submodule contains raw typings, allowing for full control (probably you don't need these unless you are implementing your own apiserver).
+//! - `Converter` together with `ConversionHandler` are middle ground: they abstract you from low-level details but not require particular conversion strategy.
+//! - `StarConverter` is opinionated strategy, where all conversions are done using intermediate unversioned representation as a middle point.
+//! Third option is preferred - this is how Kubernetes implements versioning and convertion for builtin APIs.
+
+use std::sync::Arc;
+
+use self::low_level::{ConversionResponse, ConversionReview};
+
+/// Defines low-level typings.
+pub mod low_level;
+/// High-level easy-to-use converter.
+mod star;
+
+pub use star::{StarConverter, StarConverterBuilder, StarRay};
+
+/// Helper which implements low-level API based on the provided converter.
+#[derive(Clone)]
+pub struct ConversionHandler {
+    converter: Arc<dyn Converter + Send + Sync>,
+}
+
+impl ConversionHandler {
+    /// Creates new `ConversionHandler` which will use given `converter` for operation.
+    pub fn new<C: Converter + Send + Sync + 'static>(converter: C) -> Self {
+        ConversionHandler {
+            converter: Arc::new(converter),
+        }
+    }
+
+    /// Processes `request` using stored converter so that you don't need
+    /// to think about copying `.uid` and other boring things.
+    /// Returned value is ready to be serialized and returned.
+    pub fn handle(&self, review: ConversionReview) -> ConversionReview {
+        let req = match review.request {
+            Some(r) => r,
+            None => {
+                return ConversionReview {
+                    types: review.types.clone(),
+                    request: None,
+                    response: Some(ConversionResponse::error(
+                        String::new(),
+                        ".request is unset in input".to_string(),
+                    )),
+                }
+            }
+        };
+
+        let mut converted_objects = Vec::new();
+        for (idx, object) in req.objects.into_iter().enumerate() {
+            match self.converter.convert(object, &req.desired_api_version) {
+                Ok(c) => converted_objects.push(c),
+                Err(error) => {
+                    let msg = format!("Conversion of object {} failed: {}", idx, error);
+                    return ConversionReview {
+                        types: review.types.clone(),
+                        request: None,
+                        response: Some(ConversionResponse::error(req.uid, msg)),
+                    };
+                }
+            }
+        }
```

Comment:
> Given the precedent in admission

To be honest, I think that `kube::core::admission` is too low-level and also needs some beautiful abstractions on top of it (like in kubebuilder, where you just need to implement `on_create`, `on_update` and `on_delete`) :)

> then writing this fn is no longer that daunting.

I agree that there is nothing complex in this function (especially after applying your suggestion about `.into_review`). I also agree that sometimes user will write their own handler (e.g. when conversion function uses IO, so that `async`-ness is required). However it **seems** to me that in most cases this function will suit user, so it makes some sense to have it in `kube`.

Reply:
> admission too low-level

We do actually want to stay as unopinionated as possible in the core part of kube here and represent the api, rather than make judgements on how it should be used. The runtime is really the only part that gets to make big judgements to more fit rust, otherwise deviations should be fairly small.

For admission controllers, which is so hairy to deal with in terms of cert setup anyway, I think it's probably a bigger win for the ecosystem if most users just picked or piggy-backed on something larger like [kubewarden](https://github.com/kubewarden) rather than us making two ways to do things and still forcing the full low-level cert/app pipeline on everyone.

Reply:
That said, if rust-centric, non-opinionated helper methods on those types can be made (as you said to perhaps match on various event types), in ways that does not detract from the original design, then I'm all for minor impls that help this. Admission should probably be a separate PR though. This is already large.

Reply:
> The runtime is really the only part that gets to make big judgements to more fit rust

I'm all for moving some parts of this PR (e.g. StarConversion and/or ConversionHandler) to kube_runtime. I however propose that we do these movements right before merging, after review is done.

> This is already large.

Yeah :). If this PR became in any way too big, I'm ready to split it into smaller ones (e.g. one for new structs, one for Conversion, one for StarConversion).